ipam/ipv6Ipam.go

```diff
@@ -0,0 +1,263 @@
+	defaultWindowsKubeConfigFilePath = defaultWindowsKubePath + `config`
+	defaultLinuxKubeConfigFilePath   = "/var/lib/kubelet/kubeconfig"
+	defaultSubnetMaskSizeLimit       = "/120"
+	k8sMajorVerForNewPolicyDef       = "1"
+	k8sMinorVerForNewPolicyDef       = "16"
+)
+
+// regex to get minor version
+var re = regexp.MustCompile("[0-9]+")
+
+type ipv6IpamSource struct {
+	name                string
+	nodeHostname        string
+	subnetMaskSizeLimit string
+	kubeConfigPath      string
+	kubeClient          kubernetes.Interface
+	kubeNode            *v1.Node
+	subnetRetrieved     bool
+	sink                addressConfigSink
+}
+
+// creates a new IPv6 Ipam source
+func newIPv6IpamSource(options map[string]interface{}) (*ipv6IpamSource, error) {
+	var kubeConfigPath string
+	name := options[common.OptEnvironment].(string)
+
+	if runtime.GOOS == windows {
+		kubeConfigPath = defaultWindowsKubeConfigFilePath
+	} else {
+		kubeConfigPath = defaultLinuxKubeConfigFilePath
+	}
+
+	nodeName, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+	return &ipv6IpamSource{
+		name:                name,
+		subnetMaskSizeLimit: defaultSubnetMaskSizeLimit,
+		nodeHostname:        nodeName,
+		kubeConfigPath:      kubeConfigPath,
+	}, nil
+}
+
+// Starts the MAS source.
+func (source *ipv6IpamSource) start(sink addressConfigSink) error {
+	source.sink = sink
+	return nil
+}
+
+// Stops the MAS source.
+func (source *ipv6IpamSource) stop() {
+	source.sink = nil
+}
+
+// creates a KubernetesClientset using the Kubeconfig stored on each agent node
+func (source *ipv6IpamSource) loadKubernetesConfig() (kubernetes.Interface, error) {
+	config, err := clientcmd.BuildConfigFromFlags("", source.kubeConfigPath)
+	if err != nil {
+		return nil, err
+	}
+	client, err := kubernetes.NewForConfig(config)
+
+	minimumVersion := &version.Info{
+		Major: k8sMajorVerForNewPolicyDef,
+		Minor: k8sMinorVerForNewPolicyDef,
+	}
+
+	serverVersion, err := client.ServerVersion()
+
+	isNewew := CompareK8sVer(serverVersion, minimumVersion)
+
+	if isNewew <= 0 {
+		return nil, errors.New("Incompatible Kubernetes version for dual stack")
+	}
+
+	return client, err
+}
+
+// CompareK8sVer compares two k8s versions.
+// returns -1, 0, 1 if firstVer smaller, equals, bigger than secondVer respectively.
+// returns -2 for error.
+func CompareK8sVer(firstVer *version.Info, secondVer *version.Info) int {
+	v1Minor := re.FindAllString(firstVer.Minor, -1)
+	if len(v1Minor) < 1 {
+		return -2
+	}
+	v1, err := semver.NewVersion(firstVer.Major + "." + v1Minor[0])
+	if err != nil {
+		return -2
+	}
+	v2Minor := re.FindAllString(secondVer.Minor, -1)
+	if len(v2Minor) < 1 {
+		return -2
+	}
+	v2, err := semver.NewVersion(secondVer.Major + "." + v2Minor[0])
+	if err != nil {
+		return -2
+	}
+
+	return v1.Compare(v2)
+}
+
+func (source *ipv6IpamSource) refresh() error {
+	if source == nil {
+		return errors.New("ipv6ipam is nil")
+	}
+
+	if source.subnetRetrieved {
```

Comment:
this is not persistent right? so if cni invoked again, it will query api server. I think we have to check if azure-vnet-ipam.json has the state

Reply:
where does fileIpam preserve this state?
https://github.com/Azure/azure-container-networking/blob/master/ipam/fileIpam.go#L88

Reply:
only addressmanger struct will be saved. This is not saved in state. also you can check /var/run/azure-vnet-ipam.json in one of cni clusters to check what is saved

Reply:
updated